### PR TITLE
Add parquet metadata support.

### LINKED ISSUE #1: Exposing (column) metadata (when reading Parquet files)
Is it possible to expose metadata present in Parquet files read with QuackIO? Ideally it can be attached to objects (e.g. set on a DataFrame via DataAPI metadata) that support metadata.

```
julia> df = QuackIO.read_parquet(DataFrame, "some.parquet");
julia> metadata(df)
Dict{Any, Any}()  # <-- expect metadata
```

### PR DESCRIPTION
**Pull request by evetion**
Fixes #1 

Still needs some tests

### CODE REVIEW

**Comment by aplavin** [comment]
Happy to merge when tested :)
Also, please avoid introducing unrelated whitespace changes.

**Comment by aplavin** [comment]
Does parquet support per-column metadata, or only per-table? As I understand, support for the latter is getting added here.

**Comment by evetion** [comment]
Yeah sorry for the whitespace stuff, although I'm not even sure what happened there. edit: It seems you had mixed spaces and tabs as idents (my editor converted those tabs to spaces).

This is indeed only per-table metadata, it seems Parquet doesn't support column based metadata. I'm still working on also writing metadata to file, that makes testing easier too (so we don't need to download existing parquet files).

**Comment by aplavin** [comment]
Testing on a small file created by other software would be nice anyway, no matter if you come to implementing writing or not.

> It seems you had mixed spaces and tabs as idents (my editor converted those tabs to spaces)

Ah yes, should be the results of copy the code from Pluto notebooks, they use tabs. I don't care much which of these is used - they look the same anyway – but please don't change whitespace in existing code so that not to confuse git history :)

File: src/QuackIO.jl [hunk 1]
```diff
@@ -2,13 +2,23 @@ module QuackIO
...
     con = DuckDB.DB()
-	DuckDB.register_table(con, tbl, "my_tbl")
-    qstr = "copy my_tbl to '$(escape_sql_string(file))' $(kwargs_to_db_brackets(kwargs))"
+    DuckDB.register_table(con, tbl, "my_tbl")
+
+    # Add KV_METADATA to query when metadata is supported and format is Parquet
+    qstr = if DataAPI.metadatasupport(typeof(tbl)).read && Symbol(get(kwargs, :format, nothing)) === :parquet
+        md = DataAPI.metadata(tbl)
+        kwargs = Dict{Symbol,Any}(kwargs)
+        isempty(md) || get!(kwargs, :KV_METADATA, md)
```
Comment (aplavin) [inline src/QuackIO.jl:17]: Shouldn't it be written whenever the table has metadata, even if empty?
(not too familiar with dataapi.metadata stuff)

File: src/QuackIO.jl [hunk 2]
```diff
@@ -2,13 +2,23 @@ module QuackIO
...
 export write_table, read_csv, read_parquet, read_json
 
 function write_table(file, tbl; kwargs...)
     con = DuckDB.DB()
-	DuckDB.register_table(con, tbl, "my_tbl")
-    qstr = "copy my_tbl to '$(escape_sql_string(file))' $(kwargs_to_db_brackets(kwargs))"
+    DuckDB.register_table(con, tbl, "my_tbl")
+
+    # Add KV_METADATA to query when metadata is supported and format is Parquet
+    qstr = if DataAPI.metadatasupport(typeof(tbl)).read && Symbol(get(kwargs, :format, nothing)) === :parquet
```
Comment (aplavin) [inline src/QuackIO.jl]: doesn't catch `format=:Parquet` and other capitalizations – but they are treated in the same way in duckdb

File: src/QuackIO.jl [hunk 3]
```diff
@@ -49,4 +65,14 @@ kwarg_val_to_db_incomma(x::AbstractVector) = "[" * kwarg_val_to_db(x) * "]"
 
 escape_sql_string(x::AbstractString) = replace(x, "'" => "''")
 
+function _set_parquet_metadata!(table, file)
+    qstr = """select * from parquet_kv_metadata($(kwarg_val_to_db_incomma(file)))"""
+    results = DBInterface.execute(DuckDB.DB(), qstr)
+    for (_, key, value) in results
+        skey = String(key)
+        skey == "ARROW:schema" && continue  # ignore non-string valued internal metadata
```
Comment (aplavin) [inline src/QuackIO.jl]: tests work the same with or without this line
any way to actually test that it skips `ARROW:schema`?

File: src/QuackIO.jl [hunk 4]
```diff
@@ -49,4 +65,14 @@ kwarg_val_to_db_incomma(x::AbstractVector) = "[" * kwarg_val_to_db(x) * "]"
...
+        DataAPI.metadata!(table, skey, String(value); style=:note)
```
Comment (aplavin) [inline src/QuackIO.jl]: Are values always strings, cannot they be numbers or other?

Comment (aplavin) [inline src/QuackIO.jl]: Why `style=:note`? Does it have something to do with DataAPI conventions, or with parquet format?

File: src/QuackIO.jl [hunk 5]
```diff
@@ -49,4 +65,14 @@ kwarg_val_to_db_incomma(x::AbstractVector) = "[" * kwarg_val_to_db(x) * "]"
 
 escape_sql_string(x::AbstractString) = replace(x, "'" => "''")
 
+function _set_parquet_metadata!(table, file)
+    qstr = """select * from parquet_kv_metadata($(kwarg_val_to_db_incomma(file)))"""
+    results = DBInterface.execute(DuckDB.DB(), qstr)
+    for (_, key, value) in results
```
Comment (aplavin) [inline src/QuackIO.jl]:
```suggestion
    for (;key, value) in results
```

File: src/QuackIO.jl (hunk 4 above)
Comment (aplavin) [reply src/QuackIO.jl]: DataAPI doesn't seem to define `style=:note` anywhere, the only required style is `:default`: https://github.com/JuliaData/DataAPI.jl/blob/d6d1b053c0b82e0722abf3cb1c6e6c21bbe9746d/src/DataAPI.jl#L302. I think we should just go with DataAPI recommendations then :)

Comment (aplavin) [reply src/QuackIO.jl]:
```suggestion
        DataAPI.metadata!(table, skey, String(value))
```

File: src/QuackIO.jl (hunk 1 above)
Comment (evetion) [reply src/QuackIO.jl:17]: Well, writing an empty dict (`Dict()`) doesn't do anything, and IIRC DuckDB didn't accept an empty value `{}` for `KV_METADATA`.

File: src/QuackIO.jl (hunk 2 above)
Comment (evetion) [reply src/QuackIO.jl]: Agreed, I can make a fix to always lowercase it first.

File: src/QuackIO.jl (hunk 3 above)
Comment (evetion) [reply src/QuackIO.jl:73]: I'll add a test that ARROW:schema doesn't occur in metadata.

File: src/QuackIO.jl (hunk 4 above)
Comment (evetion) [reply src/QuackIO.jl]: Hmm, I need to think about this, in fact they're just bytes (`Vector[UInt8]`) (both the key and value I believe). But that's not very useful.

File: src/QuackIO.jl (hunk 3 above)
Comment (evetion) [reply src/QuackIO.jl]: After some thought, I thought it better to leave all metadata in, and not make the choice for the user.

File: src/QuackIO.jl (hunk 4 above)
Comment (evetion) [reply src/QuackIO.jl]: After some digging, DuckDB forces both key and value to String. DataAPI requires metadata with (Abstract)String keys. So it seems logical to cast them both.

**Comment by aplavin** [comment]
Thanks! Included with some more tweaks. Should be registered soon :)

File: src/QuackIO.jl (hunk 4 above)
Comment (evetion) [reply src/QuackIO.jl]: @aplavin I missed this change, but for reference: :note seems to to be introduced by [DataFrames](https://dataframes.juliadata.org/stable/lib/metadata/), specifically for metadata that doesn't change. Default styled metadata is dropped for any operation on the object. That's good for say min/max statistics metadata, but quite bad for metadata such as author, etc. For such general metadata, :note is preferred.

Comment (aplavin) [reply src/QuackIO.jl]: I see, thanks for the explanation. Still, probably the only realistic thing we can do here is to use stuff supported by DataAPI – afaiu, it only documents the "default" style.

**Pull request merged**